ui/src/components/microsegmentation/GraphicalServiceRow.js

```diff
@@ -0,0 +1,253 @@
+import styled from '@emotion/styled';
+import { css, keyframes } from '@emotion/react';
+import ReactFlowRenderer from './ReactFlowRenderer';
+import { ReactFlowProvider } from 'react-flow-renderer';
+
+const TdStyled = styled.td`
+    background-color: ${(props) => props.color};
+    text-align: ${(props) => props.align};
+    padding: 5px 0 5px 15px;
+    vertical-align: middle;
+    word-break: break-all;
+`;
+
+const colorTransition = keyframes`
+        0% {
+            background-color: rgba(21, 192, 70, 0.20);
+        }
+        100% {
+            background-color: transparent;
+        }
+`;
+
+const TrStyled = styled.tr`
+    ${(props) =>
+        props.isSuccess &&
+        css`
+            animation: ${colorTransition} 3s ease;
+        `}
+    box-sizing: border-box;
+    margin-top: 10px;
+    box-shadow: 0 1px 4px #d9d9d9;
+    border: 1px solid #fff;
+    -webkit-border-image: none;
+    border-image: none;
+    -webkit-border-image: initial;
+    border-image: initial;
+    height: 50px;
+`;
+
+const TrStyledExpanded = styled.tr`
+    box-sizing: border-box;
+    margin-top: 10px;
+    box-shadow: 0 1px 4px #d9d9d9;
+    border: 1px solid #fff;
+    -webkit-border-image: none;
+    border-image: none;
+    -webkit-border-image: initial;
+    border-image: initial;
+    height: 50px;
+`;
+
+const LeftMarginSpan = styled.span`
+    margin-right: 10px;
+    vertical-align: bottom;
+`;
+
+const StyledDiv = styled.div`
+    padding: 10px 0 10px 0;
+    width: 100%;
+    height: 600px;
+`;
+
+export default class GraphicalServiceRow extends React.Component {
+    constructor(props) {
+        super(props);
+        this.api = props.api;
+        this.toggleView = this.toggleView.bind(this);
+        this.createElementsList = this.createElementsList.bind(this);
+        this.state = {
+            name: this.props.name,
+            errorMessage: null,
+            expandedView: false,
+        };
+    }
+
+    toggleView() {
+        if (this.state.expandedView) {
+            this.setState({
+                expandedView: false,
+            });
+        } else {
+            this.setState({
+                expandedView: true,
+            });
+        }
+    }
+
+    createElementsList() {
+        const position = { x: 0, y: 0 };
+        let newElements = [];
+        newElements.push({
+            id: '0',
+            type: 'svcNode',
+            data: {
+                data: this.props.data,
+                name: this.state.name,
+                domain: this.props.domain,
+                api: this.api,
+                _csrf: this.props._csrf,
+                onSubmit: this.props.onDeleteCondition,
+            },
+            position,
+        });
+        for (let i = 0; i < this.props.data.length; i++) {
+            let rule = this.props.data[i];
+            if (rule['category'] == 'inbound') {
+                for (let j = 0; j < rule['source_services'].length; j++) {
```

Comment:
extract this common code to a function and use it in line 150

Reply:
the code in line 150 cannot reuse this code as depending on whether the category is inbound or outbound the parameters for node and edge are different. So if I extract it to a function, I won't be able to reuse the code, if you think it will improve in readability then I can update it